File: Roman numerals/roman_numerals.cc
#include <iostream>
using namespace std;

int main() {

	int input, rest, am, x, y;
	int vals[] = {1000, 500, 100, 50, 10, 5, 1};
	char lastChar, chars[] = {'M', 'D', 'C', 'L', 'X', 'V', 'I'};
	string str;
		
	cout << "Enter a number between 1 and 3999. Enter 0 to stop.\n";

	do {
		
		cin >> input;		
		
		if (input > 0 && input < 4000 && !cin.fail()) {
			rest = input; // create duplicate so while can keep on using input
			str  = "";
			lastChar = 0;
			
			for (x = 0; x < 7; x++) {
				
				am    = rest / vals[x];
				rest %= vals[x];
				
				if (am == 4) { // maximum is 3 of same character
					if (lastChar != '\0' && ((lastChar == 'V' && x == 6) || (lastChar == 'L' && x == 4) || (lastChar == 'D' && x == 2))) { // trying to make 9__
						str = str.substr(0, (str.length() - 1)); // remove last
						str += chars[x];
						str += chars[(x-2)];
					} else { // 4__
						str += chars[x];
						str += chars[(x-1)];
					}
					lastChar = chars[x];
				} else if (am > 0) {
					for (y = 0; y < am; y++) { // add the chars
						str += chars[x];
					}
					lastChar = chars[x]; // save last char added for next loop
				}
				
			}
			
			cout << str << "\n";
		} else if (input != 0 || cin.fail()) {
			cout << "Not a valid number.\n";
			cin.clear();
			input = 1; // prevent while from stopping
		}
		
	} while (input != 0);
	
	cout << "End of program. Press Enter to close.\n";
	cin.get();
	return 0;
}

File: Roman numerals/roman_numerals.php
<?php

function roman($input) {
	$vals  = array(1000, 500, 100, 50, 10, 5, 1, 'M' => 1000, 'D' => 500, 'C' => 100, 'L' => 50, 'X' => 10, 'V' => 5, 'I' => 1);
	$chars = array('M', 'D', 'C', 'L', 'X', 'V', 'I');
	
	if (is_numeric($input) && ($input < 1 || $input > 3999)) {
		return false;
	}
	if (!is_numeric($input) && !preg_match('/[MDCLXVI]+/i', $input)) {
		return false;
	}
	
	if (is_numeric($input)) {
		$str = '';
		
		for ($x = 0; $x < 7; $x++) {
			
			$am     = floor($input / $vals[$x]);
			$input %= $vals[$x];
			
			if ($am == 4) { // max is 3 of same
				if (isset($lastChar) && (($lastChar == 'V' && $x == 6) || ($lastChar == 'L' && $x == 4) || ($lastChar == 'D' && $x == 2))) {
				 	// trying to make 9__
					$str  = substr($str, 0, -1).$chars[$x].$chars[($x-2)]; // remove last
				} else { // 4__
					$str .= $chars[$x].$chars[($x-1)];
				}
				$lastChar = $chars[$x];
			} elseif ($am > 0) {
				for ($y = 0; $y < $am; $y++) { // add the chars
					$str .= $chars[$x];
				}
				$lastChar = $chars[$x]; // save last for next loop
			}
		
		}
		return $str;
	} else {
		$input   = strtoupper($input);
		$arr     = str_split($input);
		$lastVal = 0;
		$num     = 0;
		
		foreach ($arr as $char) {
			$num += $vals[$char];
			if ($vals[$char] > $lastVal) { // trying to deduct (ex. XC -> 90)
				$num -= (2 * $lastVal); // remove added before this loop AND deduct
			}
			$lastVal = $vals[$char];
		}
		return $num;
	}
	
}

?>
